### ISSUE TITLE
Implement Q1.7 format

### ISSUE THREAD

**Issue by volcacius** [Xilinx/brevitas#12]
Reposted as a new issue. From the author:
Hi, I want to implement 8-bit fixed-point implementation of weights and activations with 7 bits for the fractional part. What value should I pass for weight_restrict_scaling_type?

_Originally posted by @MinahilRaza in https://github.com/Xilinx/brevitas/issues/4#issuecomment-552543334_

**Comment by volcacius** [comment]
For weights, you want to do the following:

```python
weight_restrict_scaling_type=RestrictValueType.POWER_OF_TWO
weight_bit_width=8,
weight_quant_type=QuantType.INT,
weight_scaling_impl_type=ScalingImplType.CONST,
weight_scaling_const=1.0  
```
This is because the actual scale factor at the end is *weight_scaling_const / 2 ^ (8 - 1)*, which is what you want for a signed Q1.7 number.

For an unsigned activation like ReLU, you need:
```python
m = 1.0
n = 7.0
max_val = 2.0 ** m - 2.0 ** - n
act = QuantReLU(
bit_width=8,
max_val=max_val,
quant_type=QuantType.INT,
restrict_scaling_type=RestrictValueType.POWER_OF_TWO,
scaling_impl_type=ScalingImplType.CONST
)
```
Reference: https://en.wikipedia.org/wiki/Q_(number_format)

Alessandro

**Comment by MinahilRaza** [comment]
I implemented the conv layer as

self.conv1 = qnn.QuantConv2d(in_channels= 1,
                                     out_channels= 20,
                                     kernel_size= 3,
                                     padding= 1,
                                     bias= False,
                                     weight_quant_type=QuantType.INT, 
                                     weight_bit_width=8,
                                     weight_restrict_scaling_type=RestrictValueType.POWER_OF_TWO,
                                     weight_scaling_impl_type=ScalingImplType.CONST,
                                     weight_scaling_const=1.0)

All of my weights should be between -1 and 1. But some of my weights are lesser that -1 ( -1.7025, -1.560 etc)
Am I missing something here?

**Comment by volcacius** [comment]
No issue on my side. An example:

```python
    def test_1_7_conv_2d():
        m = 1.0
        n = 7.0
        conv = QuantConv2d(in_channels=2,
                           out_channels=30,
                           kernel_size=1,
                           weight_bit_width=int(n + m),
                           weight_quant_type=QuantType.INT,
                           weight_restrict_scaling_type=RestrictValueType.POWER_OF_TWO,
                           weight_scaling_impl_type=ScalingImplType.CONST,
                           weight_scaling_const=1.0)

        # Init floating-point weights between -3 and -2.9 as an example
        conv.weight.data = torch.arange(start=-3, end=2.9, step=0.1)

        # Compute quant weights
        quant_weight, quant_weight_scale, quant_weight_bit_width = conv.weight_quant(conv.weight)

        # Range for Qm.n is [-(2^(m-1), 2^(m-1)-(2^(-n)) = (-1.0, 0.9922) in this case
        assert(torch.min(quant_weight) >= - (2.0 ** (m - 1.0)))
        assert(torch.max(quant_weight) <= (2.0 ** (m - 1) - 2.0 ** (- n)))

        # Resolution for Qm.n is 2^-n = (0.0078) in this case
        assert(torch.allclose(quant_weight_scale, torch.tensor(2 ** - n)))

        # Bit width is m+n
        assert(quant_weight_bit_width.item() == m+n)
```

Weights are not quantized in place, so if you call conv.weight you are still going to get the floating point version. The quantized version of the weights is computed during conv.forward(), if you want to retrieve it separately you can call conv.weight_quant(conv.weight) as i did here.
Let me know if this helps.

Alessandro

**Comment by Guyume** [comment]
@volcacius Thanks for sharing the code. I would like to ask about how to save the total quantized weight. Can you give me some detailed step?
